### PR TITLE
redesigned the pruner reconciler to fix multiple issue

### PR DESCRIPTION
**Pull request by jkandasa**
Signed-off-by: Jeeva Kandasamy <jkandasa@redhat.com>

# Changes
I have redesigned the pruner reconciler completely. the change includes fixes for the following issues,

* fixes https://github.com/tektoncd/operator/issues/1038
* fixes https://github.com/tektoncd/operator/issues/1315
* fixes https://github.com/tektoncd/operator/issues/1316
* fixes https://github.com/tektoncd/operator/issues/1317
* fixes https://github.com/tektoncd/operator/issues/1318
* fixes https://github.com/tektoncd/operator/issues/1319
* [SRVKP-2277](https://issues.redhat.com/browse/SRVKP-2277) - Prunner cronJob should run with a serviceaccount with just enough privilleges
* [SRVKP-2810](https://issues.redhat.com/browse/SRVKP-2810) - Pruner cronjob not getting updated after adding annotation operator.tekton.dev/prune.skip=true namespace

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Run `make test lint` before submitting a PR
- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes
```release-note
* Support added to https://github.com/tektoncd/operator/issues/1038 (Prune resource by pipeline).  A boolean field introduced in `TecktonConfig.spec.pruner.prune-per-resource=true` to support prune resource by `pipeline` or `taskrun`. The same thing cane be activated in a namespace by adding an annotation `operator.tekton.dev/prune.prune-per-resource=true` 
* A limited privileges service account will be created dynamically and included into pruner container.
```

### CODE REVIEW

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (piyush-garg) [inline pkg/apis/operator/v1alpha1/tektonconfig_types.go:54]: This feature is to replace the tkn image, with user-defined image in config CR, do we need this cc @vdemeester

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (piyush-garg) [inline pkg/apis/operator/v1alpha1/tektonconfig_types.go:57]: this is to apply the values of keep and keep-since at the individual resource level like if set to true, it will keep those many pipelineruns/taskruns for that pipeline/task. I think this should be good to have,  but any other name suggestions cc @vdemeester @concaf 

we will need docs to be added for this

Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:57]: I think we will need the default values also added for this as `false`

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [inline pkg/reconciler/common/prune.go:171]: This defaulting we can do at config CR level

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [inline pkg/reconciler/common/prune.go:186]: instead of creating reconcile for these three resources - serviceaccount, role binding and clusterrolebinding, we can create an installerset for them

Comment (jkandasa) [reply pkg/reconciler/common/prune.go:186]: @piyush-garg Currently I am verifying the RBAC and SA existence and owned by the operator, if not creating new entries. In near feature(after this PR merged) I am going to add a informer to react with RBAC and SA changes (deleted/modified by external user/client).
I am afraid that adding installerset may introduce bit complication to maintain prune feature. Prune is a portable feature and all the code in one location. adding installerset may introduce to work on different places and may add bit more complication on maintenance. 

WDYT?

File: pkg/reconciler/common/prune.go
Comment (jkandasa) [reply pkg/reconciler/common/prune.go:171]: @piyush-garg could you please elaborate more on this?
Would you want to add the default image in to `tektonConfig.Spec.Pruner.Image` ?

File: pkg/reconciler/common/prune.go
Comment (jkandasa) [reply pkg/reconciler/common/prune.go]: @piyush-garg I have re-implemented the RBAC/SA part. Now RBAC and ServiceAccount are managed through `TektonInstallerSet`

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [reply pkg/reconciler/common/prune.go:173]: Yes, i mean the same, if it is something specified in CR then fine, otherwise take the default from there.

here it should just read the CR value

File: hack/fetch-releases.sh
Comment (piyush-garg) [inline hack/fetch-releases.sh:278]: I dont know why we need this, we can add the config directly in kodata laso

File: pkg/reconciler/kubernetes/tektoninstallerset/controller.go
Comment (piyush-garg) [inline pkg/reconciler/kubernetes/tektoninstallerset/controller.go:71]: i am not sure why we need to watch on these

File: pkg/reconciler/shared/tektonconfig/tektonconfig.go
Comment (piyush-garg) [inline pkg/reconciler/shared/tektonconfig/tektonconfig.go:154]: if we need only if there is something specified in pruner field, otherwise not

File: hack/fetch-releases.sh
Comment (jkandasa) [reply hack/fetch-releases.sh:278]: @piyush-garg pruner feature is available for both `kubernetes` and `openshift`. adding the common `yaml` into `kodata` can lead two entries(kubernetes/openshift). Hence I added it in a common place and moved it via script.

File: pkg/reconciler/kubernetes/tektoninstallerset/controller.go
Comment (jkandasa) [reply pkg/reconciler/kubernetes/tektoninstallerset/controller.go:71]: want to react immediately, if there is change notified on the resources owned by `installerset`.
In this PR, we have introduced separate RBAC and SA for pruner. I want to reconcile them immediately, if there is a change happened.

File: pkg/reconciler/shared/tektonconfig/tektonconfig.go
Comment (jkandasa) [reply pkg/reconciler/shared/tektonconfig/tektonconfig.go:154]: we do not have separate CR for pruner. still it is possible by having pruner section hash value.

I do not know how to track, if pruner RBAC/SA `installerset` get deleted. hence reconciling on each call.

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (jkandasa) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:54]: @vdemeester could you please provide your thoughts on this?

File: pkg/reconciler/kubernetes/tektoninstallerset/controller.go
Comment (piyush-garg) [reply pkg/reconciler/kubernetes/tektoninstallerset/controller.go:71]: i dont think it is needed, we dont need to react on manual changes immediately, we can wait for next reconcile

File: pkg/reconciler/shared/tektonconfig/tektonconfig.go
Comment (piyush-garg) [reply pkg/reconciler/shared/tektonconfig/tektonconfig.go:154]: we can ccheck if config. spec.pruner is not nil

File: pkg/reconciler/kubernetes/tektoninstallerset/controller.go
Comment (jkandasa) [reply pkg/reconciler/kubernetes/tektoninstallerset/controller.go:71]: IMO, the purpose of the operator should keep the desired state of the application all the time as soon as possible.
RBAC/SA needs for `pruner` job. pruner job will fail if something not found.

File: pkg/reconciler/shared/tektonconfig/tektonconfig.go
Comment (jkandasa) [reply pkg/reconciler/shared/tektonconfig/tektonconfig.go:154]: Sure, I will update the code to skip pruner `installerset` reconcile, `if config. spec.pruner == nil`

Comment (jkandasa) [reply pkg/reconciler/shared/tektonconfig/tektonconfig.go:154]: @piyush-garg I have updated the code to skip reconcile pruner installerset

File: pkg/reconciler/kubernetes/tektoninstallerset/controller.go
Comment (jkandasa) [reply pkg/reconciler/kubernetes/tektoninstallerset/controller.go]: @piyush-garg I have removed this change from this PR. I have handled this via https://github.com/tektoncd/operator/pull/1442

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (jkandasa) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:57]: @vdemeester @piyush-garg  How about 
* `KeepByResource` => `keep-by-resource`
* `LimitByResource` => `limit-by-resource`
* `FilterByResource` => `filter-by-resource`

I am poor at naming, please suggest a name :slightly_smiling_face: 

Note: we are limiting the number of resources with `keep` parameter

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (jkandasa) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:54]: @vdemeester I am going to remove this image option, we can manage it via environment variable.
* we can update it via operator deployment environment
* in olm managed version, we can update the environment in subscription cr

File: pkg/reconciler/common/prune.go
Comment (jkandasa) [reply pkg/reconciler/common/prune.go]: @piyush-garg I reverted the image option via CR

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (piyush-garg) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go:54]: I think `prune-per-resource` is better

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [inline pkg/reconciler/common/prune.go:173]: this is not required right

if there is an env, we set that image. If there is no env, we set the default whatever is there in the cronjob

Comment (piyush-garg) [reply pkg/reconciler/common/prune.go:173]: we can just add that as tranformer, no need to do validation for this

Comment (jkandasa) [reply pkg/reconciler/common/prune.go:173]: @piyush-garg We need this. This is the only place we are getting the image detail and adding same into CronJob. there is no default image for pruner. It is mandatory to have environment `IMAGE_JOB_PRUNER_TKN` (prunerContainerImageEnvKey)

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [inline pkg/reconciler/common/prune.go:366]: is not better to move this check before normalization

File: pkg/reconciler/common/prune.go
Comment (piyush-garg) [inline pkg/reconciler/common/prune.go:353]: is not better to do validation in config_validation.go

File: pkg/reconciler/shared/tektonconfig/installerset_pruner.go
Comment (jkandasa) [inline pkg/reconciler/shared/tektonconfig/installerset_pruner.go:175]: check the transformer for changing the namespace for openshift

File: hack/fetch-releases.sh
Comment (jkandasa) [inline hack/fetch-releases.sh:232]: act on make clean

File: docs/TektonConfig.md
Comment (jkandasa) [inline docs/TektonConfig.md:152]: @piyush-garg kindly review the doc text

File: hack/fetch-releases.sh
Comment (jkandasa) [reply hack/fetch-releases.sh:232]: @piyush-garg `make clean` executes `rm -rf ./cmd/kubernetes/operator/kodata/tekton*` 
so no action needed here.

File: pkg/reconciler/common/transformers.go
Comment (jkandasa) [inline pkg/reconciler/common/transformers.go:851]: @piyush-garg I have added these transformers to replace namespace in SA and ClusterRoleBinding. I could not find existing func for this. Please let me know if there is already available. If not I can continue with this and I will a unit test for theses functions

Comment (jkandasa) [reply pkg/reconciler/common/transformers.go:851]: unit test added

File: pkg/apis/operator/v1alpha1/tektonconfig_types.go
Comment (jkandasa) [reply pkg/apis/operator/v1alpha1/tektonconfig_types.go]: @piyush-garg updated

File: pkg/reconciler/common/transformers.go
Comment (piyush-garg) [reply pkg/reconciler/common/transformers.go:851]: there are tranformers already

**Comment by piyush-garg** [comment]
cc @vdemeester

**Comment by jkandasa** [comment]
@piyush-garg rebase done, as need. you have to `lgtm` label again. thanks!

**Review by vdemeester — APPROVED** [review]

**Pull request merged**
